containers/init/ready/ready.go

```diff
@@ -324,13 +375,91 @@ func main() {
 		log.Fatalf("failed to wait for ready pods: %v", err)
 	}
 
-	log.Printf("all pods ready, exiting successfully")
+	log.Printf("all pods ready")
 	workerFileBody := strings.Join(podIPs, ",")
 	ioutil.WriteFile(outputFile, []byte(workerFileBody), 0777)
 
 	test, err := grpcClientset.LoadTestV1().LoadTests(corev1.NamespaceDefault).Get(ctx, os.Args[1], metav1.GetOptions{})
 	if err != nil {
-		log.Fatalf("failed to fetch loadtest for obtaining metadata: %v", err)
+		log.Fatalf("failed to fetch loadtest: %v", err)
+	}
+
+	if clientSpecValid, err := kubehelpers.IsClientsSpecValid(&test.Spec.Clients); !clientSpecValid {
+		log.Fatalf("validation failed in checking clients' spec: %v", err)
+	}
+
+	if kubehelpers.IsPSMTest(&test.Spec.Clients) {
+		log.Printf("running PSM test, prepare to send backends information and test type to xds server")
+
+		endpoints := buildEndpoints(nodesInfo.Servers, uint32(testconfig.TestServerPort))
+
+		isProxied := kubehelpers.IsProxiedTest(&test.Spec.Clients)
+		if isProxied {
+			log.Printf("running proxied test")
+		} else {
+			log.Printf("running proxyless test")
+		}
+
+		psmTargetOverride := ""
+		for _, clientNode := range nodesInfo.Clients {
+
+			currentPSMTargetOverride, err := communicateWithEachClient(clientNode.PodIP, endpoints, isProxied)
+			if err != nil {
+				log.Fatalf("failed to communicate backend endpoints to client %v: %v", clientNode.Name, err)
+			}
+
+			if psmTargetOverride == "" {
+				psmTargetOverride = currentPSMTargetOverride
+			} else {
+				if psmTargetOverride != currentPSMTargetOverride {
+					log.Fatalf("not all client have the same server target")
+				}
+			}
+		}
+
+		if psmTargetOverride != "" {
+			if err := ioutil.WriteFile(DefaultServerTargetOverrideOutputFile, []byte(psmTargetOverride), 0777); err != nil {
+				log.Fatalf("failed to write PSM server target: %v to %v", psmTargetOverride, DefaultServerTargetOverrideOutputFile)
+			}
+			log.Printf("write PSM server target: %v to %v\n", psmTargetOverride, DefaultServerTargetOverrideOutputFile)
+		} else {
+			log.Fatalf("failed to obtain PSM server target")
+		}
+
+		if isProxied {
+			envoyListenerPort := strings.Split(psmTargetOverride, ":")[1]
```

Comment:
Is this string validated anywhere? I'm wondering about an out-of-bounds access.

Reply:
Added a check here.